Feature: parse messages as json

ThisIsMissEm opened the pull request:
# New feature description

- Parses the websocket messages as JSON, instead of leaving this to the end-user
- Correctly documents events emitted from the WebsocketNotification class
- Switch to `getWellKnownSolid` from `solid-client` to work across both ESS 1.1 and ESS 1.2
- Fix unit tests (migrate from jest-fetch-mock which caused issues for e2e tests)
- Fix & re-enable e2e browser and node.js tests
- Drop support for Node.js v12 (end of life)
- Migrate CI/CD to use new environments (ESS PodSpaces & ESS PodSpaces Next)

# Checklist

- [x] All acceptance criteria are met.
- [x] Relevant documentation, if any, has been written/updated.
- [x] The changelog has been updated, if applicable.
- [ ] New functions/types have been exported in `index.ts`, if applicable. n/a
- [ ] New modules (i.e. new `.ts` files) are listed in the `exports` field in `package.json`, if applicable. n/a
- [ ] New modules (i.e. new `.ts` files) are listed in the `typedocOptions.entryPoints` field in `tsconfig.json`, if applicable. n/a
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

ThisIsMissEm committed:
feature: parse websocket messages as JSON

kay-kim commented on docs/api/source/index.rst:
```diff
@@ -13,7 +13,24 @@ For Solid Servers that support notifications, they are sent when a resource is
+   future release will support this draft specification once implementations
+   become available. 
+   
+   **Please note:** the `legacy (insecure) WebSocket
```
So ... I would refrain from labeling this as "insecure" because then people will ask you to enumerate in all ways you mean by insecure, etc. 
But will let @nicolasmondada  weigh in.

ThisIsMissEm replied:
This wording largely came from @acoburn — perhaps he can give some light on this.

nicolasmondada replied:
Let's avoid using nouns on things we don't recommend support and be specific that "Inrupt" does not recommend it. Let's just say: "The legacy Websocket API, which is not recommended by Inrupt for secure messaging, is not supported  by this library

ThisIsMissEm replied:
The comment on security: https://github.com/solid/solid-spec/blob/master/api-websockets.md#status

kay-kim commented on docs/api/source/index.rst:
```diff
@@ -13,7 +13,24 @@ For Solid Servers that support notifications, they are sent when a resource is
+   <https://docs.inrupt.com/ess/1.1/services/service-websocket/>`__, which
+   formed the basis of the Solid Notification Protocol 1.0 draft specification.
+
+   The `Solid Notification Protocol 1.0
```
I'm not quite sure what this paragraph is adding, since this library doesn't support the draft spec.  
Aside:  as a general rule of thumb, we try to avoid forward-looking statements in our docs. 

If you want to say that it's subject to change, that's okay.

ThisIsMissEm replied:
Yeah, that's the intent: there's a newer protocol, we'll be hopefully implementing it soon, but we don't currently implement it.

nicolasmondada replied:
Please lets avoid making any roadmap statements in here.

kay-kim commented on docs/api/source/index.rst:
```diff
@@ -13,7 +13,24 @@ For Solid Servers that support notifications, they are sent when a resource is
+
+   This library currently only supports an `early version of WebSocket
+   notifications
+   <https://docs.inrupt.com/ess/1.1/services/service-websocket/>`__, which
```
So ... I might be a bit more explicit in that it supports ESS WebSocket notifications. 
And omit all about forming the basis of the Solid Notification Protocol ...

ThisIsMissEm replied:
I'll defer to @acoburn here.

nicolasmondada replied:
Agree with Kay here. Let's just say "This library currently only supports ESS Websocket notifications."

ThisIsMissEm committed:
fix: typescript errors in node.js e2e tests

kay-kim reviewed (commented):
Hey @ThisIsMissEm -  just left a little comment about the landing page blurbs.  You might want to work with @nicolasmondada  about how to position the product.

ThisIsMissEm committed:
fix: node.js end-to-end tests

ThisIsMissEm committed:
fix: browser end-to-end tests

ThisIsMissEm committed:
chore: upgrade playwright

ThisIsMissEm committed:
fix: ensure e2e browser test-results are outputted to the correct directory

ThisIsMissEm committed:
fix: e2e browser tests need to execute the login steps in series, not parallel

ThisIsMissEm committed:
fix: wait for navigation when setting up browser e2e tests

ThisIsMissEm committed:
fix: ensure dotenv is only loaded on developer machines & not on CI

ThisIsMissEm committed:
fix: use getWellKnownSolid from solid-client

NSeydoux commented on e2e/browser/test/roles.ts:
```diff
@@ -33,10 +33,10 @@ export const essUserLogin = async (
-    authorisePage.authoriseOnce(),
-    indexPage.handleRedirect(),
-  ]);
+  // Note: these steps must execute in series, not parallel, which is what
+  // Promise.all would do:
+  await indexPage.startLogin();
+  await cognitoPage.login(login, password);
+  await authorisePage.authoriseOnce();
+  await indexPage.handleRedirect();
```
That makes total sense, and I'm surprised this hasn't been an issue before.

nicolasmondada commented on docs/api/source/index.rst:
```diff
@@ -13,7 +13,24 @@ For Solid Servers that support notifications, they are sent when a resource is
+   future release will support this draft specification once implementations
+   become available. 
+   
+   **Please note:** the `legacy (insecure) WebSocket
```
```suggestion
   **Please note:** the `legacy WebSocket
```

nicolasmondada commented on docs/api/source/index.rst:
```diff
@@ -13,7 +13,24 @@ For Solid Servers that support notifications, they are sent when a resource is
+   
+   **Please note:** the `legacy (insecure) WebSocket
+   <https://github.com/solid/solid-spec/blob/master/api-websockets.md>`__
+   mechanism, which is not recommended for secure messaging, is not supported by
```
```suggestion
   API, which is not recommended by Inrupt for secure messaging, is not supported by
```

ThisIsMissEm committed:
refactor: make e2e node tests able to execute in parallel

This is based on the strategy from solid-client-js

ThisIsMissEm committed:
refactor: migrate away from jest-fetch-mock

It turned out to be buggy and resulted in the e2e tests failing weirdly as it somehow was always loaded at a typescript level

ThisIsMissEm committed:
refactor: make getDefaultSessionFetch more robust

Previously there was potential for a race-condition to happen due to the asynchronous loading of @inrupt/solid-client-authn-browser

ThisIsMissEm committed:
refactor: cleanup e2e workflows

ThisIsMissEm committed:
refactor: test on PodSpaces Next

ThisIsMissEm committed:
refactor: use new environments

ThisIsMissEm committed:
chore: disable E2E tests against PodSpaces until getWellKnownSolid works on 1.1

ThisIsMissEm committed:
chore: drop support for Node.js v12.x

ThisIsMissEm committed:
chore: test by default against current LTS Node.js version (v16.x)

ThisIsMissEm committed:
chore: add default publishConfig setting to package.json

ThisIsMissEm committed:
fix: make LiveNotification#emitter hidden, not internal

This allows us to use the emitter in e2e tests, until we have once/off supported

ThisIsMissEm committed:
fix: continuous deployment tests

ThisIsMissEm committed:
fixup! fix: continuous deployment tests

ThisIsMissEm committed:
fix: disable cd-end-to-end-tests, as it's flakey and doesn't actually give us any value

ThisIsMissEm committed:
fix: release workflow for new environments

ThisIsMissEm commented on .github/workflows/cd.yml:
```diff
@@ -216,46 +217,44 @@ jobs:
-          E2E_TEST_CLIENT_ID: ${{ secrets.E2E_TEST_CLIENT_ID }}
-          E2E_TEST_CLIENT_SECRET: ${{ secrets.E2E_TEST_CLIENT_SECRET }}
-          E2E_TEST_ENVIRONMENT: "ESS Production"
+  # cd-end-to-end-tests:
```
This was broken for some reason and I couldn't get it working for the life of me. This is probably not important.

ThisIsMissEm commented on CHANGELOG.md:
```diff
@@ -6,6 +6,14 @@ The following changes have been implemented but not released yet:
+### New Features
+
+- WebsocketNotification `message` events are now parsed as JSON, this is technically a breaking change but we're allowing it due to the alpha nature of this library.
+
```
```suggestion

### Fixes

- use `getWellKnownSolid` from `@inrupt/solid-client` to be more compatible/reliable
- improve documentation of entire package to be clearer and document only APIs that should be used by end-users of the SDK.
- rework all tests to run successfully
```

ThisIsMissEm committed:
Update CHANGELOG.md

solid-akb commented on e2e/node/e2e.test.ts:
```diff
@@ -33,156 +33,124 @@ import {
-const testIf = (condition: boolean) => (condition ? it : it.skip);
+const nextWebsocketMessage = async (ws: WebsocketNotification) => {
+  return new Promise((resolve) => {
+    // TODO: implement ws.once
```
do we need this comment still?

ThisIsMissEm replied:
Yeah, we've a ticket on the backlog for it, as #246 starts to implement it; really we need this to clean up receiving messages for expectations, the way we're doing it now is very hacky/problematic.

ThisIsMissEm commented on e2e/browser/test/e2e.playwright.ts:
```diff
@@ -30,14 +30,14 @@ import {
 
 const { login, password, notificationGateway } = getTestingEnvironmentBrowser();
 
 // FIXME: There is currently a bug where the websocket is created, but no events
```
Not sure; I think I'd want to leave the comment in for now, and then remove it later; there must have been a reason this was here. cc @NSeydoux

ThisIsMissEm committed:
tests: cleanup old comments about flaky tests
